lib/sycamore/sycamore/utils/derivation_utils.py

```diff
@@ -0,0 +1,202 @@
+            return False  # Don't overwrite existing values
+
+        from_value = self.properties.get(from_property_name)
+        if from_value is None:
+            return False
+
+        if to_property_name in self.unit_map and from_property_name in self.unit_map:
+            to_unit = self.unit_map[to_property_name]
+            from_unit = self.unit_map[from_property_name]
+
+            try:
+                converted_value = self.unitConverter.convert(from_value, from_unit, to_unit)
+                self.properties[to_property_name] = converted_value
+                self.metadata[to_property_name] = [from_property_name]
+                return True
+            except Exception:
+                return False
+        return False
+
+    def derive_all_conversion(self, to_property_name, from_property_name, unit_map: None):
+        self.unit_map = unit_map or self.unit_map
+        if to_property_name not in self.properties and from_property_name not in self.properties:
+            return False
+        value1 = self.properties.get(from_property_name)
+        if value1 is None:
+            return False
+        if to_property_name in self.unit_map and from_property_name in self.unit_map:
+            to_unit = self.unit_map[to_property_name]
+            from_unit = self.unit_map[from_property_name]
+            if value1 is not None:
+                converted_value = self.unitConverter.convert(value1, from_unit, to_unit)
+                self.properties[to_property_name] = converted_value
+                self.metadata[to_property_name] = [from_property_name]
+                return True
+        return False
+
+    def fill_missing_units(self, list_of_fields, unit_map):
+        fields_with_value = [field for field in list_of_fields if self.properties.get(field) is not None]
+        if len(fields_with_value) == 0:
+            return
+        reference_field = fields_with_value[0]
+        for field in list_of_fields:
+            if field == reference_field:
+                continue
+
+            if self.properties.get(field) is None:
+                self.derive_all_conversion(field, reference_field, unit_map)
+
+
+class PropertyDerivation(UnitConverter, UnitDerivation):
+    def __init__(self, properties, metadata={}, unit_map={}):
+        self.properties = properties
+        self.metadata = metadata
+        UnitConverter.__init__(self)
+        self.unit_map = unit_map
+        UnitDerivation.__init__(self, self.properties, self.metadata, self, self.unit_map)
+        self.group = []
+
+    def unit_group(self, group):
+        temp_group = []
+        for property in group:
+            temp_group.append(property)
+        self.fill_missing_units(temp_group, self.unit_map)
+        self.group.append(temp_group)
+
+    def fill_from_formula(self, formula_string: str):
+        ops = "+-*/"
+        left, right = [part.strip() for part in formula_string.split("=")]
+        if any(op in left for op in ops):
```

Comment:
I'd just force people to write the second form (target = a op b) and assert the condition.